src/ok.rs

```diff
@@ -0,0 +1,125 @@
+use std::str::from_utf8;
+
+#[derive(StructOpt)]
+pub enum SevGeneration {
+    #[structopt(about = "AMD SEV Extensions")]
+    Sev,
+
+    #[structopt(about = "AMD SEV-ES Extensions")]
+    Es,
+}
+
+struct CpuId {
+    pub name: &'static str,
+    pub leaf: u32,
+    pub func: fn(x86_64::CpuidResult) -> (bool, Option<String>),
+}
+
+struct TestResult {
+    pub success: bool,
+    pub msg: String,
+}
+
+const CORE_CPUID: &[CpuId] = &[CpuId {
+    name: "CPU Manufacturer",
+    leaf: 0x00000000,
+    func: |res| {
+        let name: [u8; 12] = unsafe { transmute([res.ebx, res.edx, res.ecx]) };
+        let name = from_utf8(&name[..]).unwrap();
+
+        (name == "AuthenticAMD", None)
+    },
+}];
+
+const AMD_CPU_DEPENDENT_CPUIDS: &[CpuId] = &[
+    CpuId {
+        name: "Microcode support",
+        leaf: 0x80000002,
+        func: |_res| {
+            let cpu_name = {
+                let mut bytestr = Vec::with_capacity(48);
+                for cpuid in 0x8000_0002_u32..=0x8000_0004_u32 {
+                    let cpuid = unsafe { x86_64::__cpuid(cpuid) };
+                    let mut bytes: Vec<u8> = [cpuid.eax, cpuid.ebx, cpuid.ecx, cpuid.edx]
+                        .iter()
+                        .flat_map(|r| r.to_le_bytes().to_vec())
+                        .collect();
+                    bytestr.append(&mut bytes);
+                }
+                String::from_utf8(bytestr).unwrap().trim().to_string()
+            };
+
+            (cpu_name.to_uppercase().contains("EPYC"), None)
+        },
+    },
+    CpuId {
+        name: "Secure Memory Encryption (SME)",
+        leaf: 0x8000001f,
+        func: |res| (res.eax & 0x1 != 0, None),
+    },
+];
+
+const MICROCODE_SME_DEPENDENT_CPUIDS: &[CpuId] = &[CpuId {
+    name: "Secure Encrypted Virtualization (SEV)",
+    leaf: 0x8000001f,
+    func: |res| (res.eax & (0x1 << 1) != 0, None),
+}];
+
+const SEV_DEPENDENT_CPUIDS: &[CpuId] = &[
+    CpuId {
+        name: "Page flush MSR",
+        leaf: 0x8000001f,
+        func: |res| (res.eax & (0x1 << 2) != 0, None),
+    },
+    CpuId {
+        name: "Physical address bit reduction",
+        leaf: 0x8000001f,
+        func: |res| {
+            let field = res.ebx & 0b1111_1100_0000 >> 6;
+            (true, Some(format!("{}", field)))
+        },
+    },
+    CpuId {
+        name: "C-bit location in page table entry",
+        leaf: 0x8000001f,
+        func: |res| {
+            let field = res.ebx & 0b01_1111;
+            (true, Some(format!("{}", field)))
+        },
+    },
+    CpuId {
+        name: "Number of encrypted guests supported simultaneously",
+        leaf: 0x8000001f,
+        func: |res| (true, Some(format!("{}", res.ecx))),
+    },
+    CpuId {
+        name: "Minimum ASID value for SEV-enabled, SEV-ES disabled guest",
+        leaf: 0x8000001f,
+        func: |res| (true, Some(format!("{}", res.edx))),
+    },
+];
+
+// SEV-ES specific CPUIDs.
+const SEV_ES_CPUIDS: &[CpuId; 1] = &[CpuId {
```

Comment:
Probably doesn't need to be an array with a specified length, can just be a `&[Cpuid]` can't it?

Reply:
This should be fixed in this commit, not a later commit.